service/raft/raft_group0.cc

```diff
@@ -402,13 +402,15 @@ future<> raft_group0::leadership_monitor_fiber() {
             }
         });
 
+        bool was_leader = group0_server().is_leader();
         while (true) {
-            while (!group0_server().is_leader()) {
-                co_await group0_server().wait_for_state_change(&_leadership_monitor_as);
-            }
-            group0_log.info("gaining leadership");
             co_await group0_server().wait_for_state_change(&_leadership_monitor_as);
-            group0_log.info("losing leadership");
+            if (was_leader) {
+                group0_log.info("losing leadership");
+            } else if (group0_server().is_leader()) {
+                group0_log.info("gaining leadership");
+            }
```

Comment:
I'm not sure this is correct.
Suppose that the server was a leader at the beginning of the loop. We `wait_for_state_change`. Then the server loses leadership and quickly gains it back, before `wait_for_state_change` wakes up.

In that case we will print `losing_leadership` and won't print `gaining_leadership`, then enter another wait.

The previous version didn't have this problem.

Is printing "gaining leadership" after being chosen discovery leader really a problem? I wouldn't say. I think it's pretty consistent -- the server always prints "gaining leadership" when becoming a group 0 leader, irrespective of why it happened.

Reply:
The problem is that "gaining leadership" is not printed instantly after being chosen as the discovery leader because `leadership_monitor_fiber` is not even started. A mistimed "gaining leadership" log might be confusing. But even if it is a problem, it's much smaller than the incorrectness described by you. I'll remove the second commit.